### ISSUE TITLE
`consistent-destructuring` improvements

### ISSUE THREAD

**Issue by zloirock** [sindresorhus/eslint-plugin-unicorn#1005]
It could be good to add an option for ignoring redefined properties of sources:
```js
// I don't see a reason to use destructuring for `edge`
const { chrome, ie, safari } = module;
// ...
if (!module.edge && ie) {
  module.edge = '12';
}
// ...
if (!module.edge) {
  // ...
}
```
It could be good to add an option to ignore some sources or properties:
```js
const { getPrototypeOf } = Object;
// ...
// `Object.prototype` here looks better than just `prototype`
if (getPrototypeOf(object) === Object.prototype) {
  // ...
}
```

**Comment by brettz9** [comment]
Another improvement, imo, would be to allow referencing a higher property that was deeply destructured:

```js
const {request: {cache, mode, url}} = e;
doSomething(cache, mode, url);

doAnother(e.request);
```

**Comment by GuoSirius** [comment]
Another improvement, aim at Vue@3.0，It's no longer responsive when you deconstruct it

```javascript
<script>
import { defineComponent, watch } from 'vue'

export default defineComponent({
  name: 'Hello',
  setup(props, context) {
    const { a, b, c } = props

    if (a && b && c) {
      // TODO
    }

    watch(() => props.a, () => {})
    watch(() => props.d, () => {})
  }
})
</script>
```

**Comment by GuoSirius** [comment]
```javascript
asyncRequest(result => {
  const { a, b, c, d = [] } = result

  const e = result.e || []

  // The value of d and e is different
})
```

**Comment by sindresorhus** [comment]
// @medusalix @nickofthyme @MrHen @futpib Mentioning you from the original issue for the rule in case you have any thoughts on these possible improvements.

**Comment by fisker** [comment]
// @zloirock @brettz9  @GuoSirius Can you take a look my new proposal https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1230, I think this is more like you want.

**Comment by brettz9** [comment]
Regarding your Vue case, @GuoSirius , I think that could be met by https://eslint.vuejs.org/rules/no-setup-props-destructure.html .

**Comment by futpib** [comment]
I think examples provided here can be rewritten in a better way that does conform to the latest (31.0.0) unicorn rules:

@zloirock's from https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1005#issue-783525204:
```js
let {chrome, ie, safari, edge} = module;

if (!edge && ie) {
	edge = '12';
}

if (!edge) {
	// ...
}

// If you *really* need to mutate `module.edge` for some reason,
// do it after you changed all destructured variables you wanted to change:
Object.assign(module, {
	chrome,
	ie,
	safari,
	edge,
});
```
```js
if (Object.getPrototypeOf(object) === Object.prototype) {
	// ...
}
```

@brettz9's from https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1005#issuecomment-758325847

This already passes unicorn rules, but you could rewrite it to use even more destructuring with IMO even better readability:
```js
const {request} = event;
const {cache, mode, url} = request;

doSomething(cache, mode, url);
doAnother(request);
```

I don't get exactly which improvements are proposed here. I think this needs more clear examples with comments like "This line is reported (and/or autofixed) but I think it should not be reported" or "This line is not reported but I think it should be reported (and/or autofixed)" each in a separate issue.

**Comment by brettz9** [comment]
Yes, they *can* be redone, but some of us consider it troublesome to have to refactor, e.g., changing `const` to `let`, just to get it to work. Also, in my case, I'm aware I could make those steps, but I don't tend to wish to add lines for just a one-off use (and `unicorn/consistent-destructuring` *does* report this).

**Comment by brettz9** [comment]
I might just add that I think there are essentially three approaches that are favored, none of which is necessarily superior, but the current rule on enforces one of them.

1. Require it everywhere possible (the existing rule)
2. Require it wherever a destructuring structure makes it available (@fisker's proposal)
3. Require it wherever a destructuring structure currently makes it feasible (my own preference)

By my own preference, I mean the following:

```js
const {a, b, c: {d}} = obj

obj.c // Should *not* report did it would require retooling (a new line)
if (!obj.f) { // Should *not* report as it would require retooling (replacing with `let` or a `let` and `const`)
  obj.f = '12';
}
obj.e // Should report since it can just be added to the destructuring
```

**Comment by sindresorhus** [comment]
This should now be covered by the simplified `consistent-destructuring` behavior: the rule only reports properties that already have a corresponding destructured variable, and it no longer tries to expand existing destructuring patterns.

I added regression coverage for the examples in this issue and documented that there is no ignore option because properties without a corresponding destructured variable are ignored by design.
